src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -1529,6 +1533,24 @@ class Checker(globalTable: SymbolTable, verbose: Boolean = false) {
 
     }
 
+    private def checkConstructors(constructors: Seq[Constructor], contract: Contract, table: ContractTable): Unit = {
+        if (constructors.isEmpty && table.stateLookup.nonEmpty) {
+            logError(contract, NoConstructorError(contract.name))
+        }
+
+        if (constructors.length > 1 && contract.isMain) {
+            logError(contract, MultipleConstructorsError(contract.name))
+        }
+
+        val constructorsByArgTypes = constructors.groupBy(c => c.args.map(_.typ))
+        val matchingConstructors = constructorsByArgTypes.filter(_._2.size > 1)
+
+        matchingConstructors.foreach(typeAndConstructors => {
+            val highestLine = typeAndConstructors._2.maxBy(_.loc.line)
```

Comment:
Neat way of doing this. But "highest" is confusing because that might mean the line with lower line number. Consider "greatest" instead.